connect_ext_test.go

```diff
@@ -2968,6 +2968,379 @@ func TestSetProtocolHeaders(t *testing.T) {
+						break
+					}
+					return err
+				}
+				if req.GetNumber() == -99 {
+					// Special case: error after successful exchanges (for testing trailers)
+					callInfo.ResponseHeader().Set("x-custom-key", "cumsum-error-streaming")
+					callInfo.ResponseHeader().Set("x-header-only", "should-not-be-in-trailers")
+					callInfo.ResponseTrailer().Set("x-trailer-only", "should-not-be-in-headers")
+					callInfo.ResponseTrailer().Set("x-both-sources", "from-callinfo-trailer")
+					err := connect.NewError(connect.CodeInternal, nil)
+					err.Meta().Set("x-error-only", "from-error-only")
+					err.Meta().Set("x-both-sources", "from-error-meta")
+					return err
+				}
+				if req.GetNumber() < 0 {
+					callInfo.ResponseHeader().Set("x-custom-key", "cumsum-error")
+					callInfo.ResponseHeader().Set("x-header-only", "should-not-be-in-trailers")
+					callInfo.ResponseTrailer().Set("x-trailer-only", "should-not-be-in-headers")
+					callInfo.ResponseTrailer().Set("x-both-sources", "from-callinfo-trailer")
+					err := connect.NewError(connect.CodeInvalidArgument, nil)
+					err.Meta().Set("x-error-only", "from-error-only")
+					err.Meta().Set("x-both-sources", "from-error-meta")
+					return err
+				}
+				sum += req.GetNumber()
+				if err := stream.Send(&pingv1.CumSumResponse{Sum: sum}); err != nil {
+					return err
+				}
+			}
+			return nil
+		},
+	}
+
+	mux := http.NewServeMux()
+	mux.Handle(pingv1connectsimple.NewPingServiceHandler(handler))
+	server := memhttptest.NewServer(t, mux)
+
+	testCallInfoHeaders := func(t *testing.T, client pingv1connectsimple.PingServiceClient, protocol string) { //nolint:thelper
+		t.Run("unary_ping_success", func(t *testing.T) {
+			t.Parallel()
+			ctx, callInfo := connect.NewClientContext(t.Context())
+			response, err := client.Ping(ctx, &pingv1.PingRequest{Number: 1})
+			assert.Nil(t, err)
+			assert.NotNil(t, response)
+
+			headerValue := callInfo.ResponseHeader().Get("x-custom-key")
+			assert.Equal(t, headerValue, "ping-success")
+			assert.Equal(t, callInfo.ResponseHeader().Get("x-success-header"), "in-headers")
+			assert.Equal(t, callInfo.ResponseTrailer().Get("x-success-trailer"), "in-trailers")
+			assert.Equal(t, callInfo.ResponseTrailer().Get("x-success-header"), "")
+			assert.Equal(t, callInfo.ResponseHeader().Get("x-success-trailer"), "")
+		})
+		t.Run("unary_ping_error", func(t *testing.T) {
+			t.Parallel()
+			ctx, callInfo := connect.NewClientContext(t.Context())
+			response, err := client.Ping(ctx, &pingv1.PingRequest{Number: -1})
+			assert.NotNil(t, err)
+			assert.Nil(t, response)
+
+			headerValue := callInfo.ResponseHeader().Get("x-custom-key")
+			assert.Equal(t, headerValue, "ping-error")
+			assert.Equal(t, callInfo.ResponseHeader().Get("x-header-only"), "should-not-be-in-trailers")
+			assert.Equal(t, callInfo.ResponseTrailer().Get("x-trailer-only"), "should-not-be-in-headers")
+			assert.Equal(t, callInfo.ResponseHeader().Get("x-trailer-only"), "")
```

Comment:
using `http.Header.Get()` seems risky because it will succeed, even if the headers actually include more than one value. I think you need to use `Values()` and then update expectation to be a single-element slice literal.

Reply:
Assert the values now.